br/pkg/checksum/executor.go

```diff
@@ -0,0 +1,321 @@
+		}
+		peer = regionInfo.Region.Peers[0]
+	}
+	storeID := peer.GetStoreId()
+	store, err := splitClient.GetStore(ctx, storeID)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	conn, err := grpc.Dial(store.GetAddress(), grpc.WithInsecure())
+	if err != nil {
+		return errors.Trace(err)
+	}
+	defer conn.Close()
+
+	client := tikvpb.NewTikvClient(conn)
+	rangeStart, rangeEnd := adjustRegionRange(exec.startKey, exec.endKey, regionInfo.Region.StartKey, regionInfo.Region.EndKey)
+	curStart := rangeStart
+	firstLoop := true
+	checksum := Checksum{}
+	digest := crc64.New(crc64.MakeTable(crc64.ECMA))
+	apiver := exec.apiVersion
+	if apiver == kvrpcpb.APIVersion_V1TTL {
+		apiver = kvrpcpb.APIVersion_V1
+	}
+	kvrpcpbCtx := &kvrpcpb.Context{
+		RegionId:    regionInfo.Region.Id,
+		RegionEpoch: regionInfo.Region.RegionEpoch,
+		Peer:        peer,
+		ApiVersion:  apiver,
+	}
+	for {
+		resp, err := client.RawScan(ctx, &kvrpcpb.RawScanRequest{
+			Context:  kvrpcpbCtx,
+			StartKey: curStart,
+			EndKey:   rangeEnd,
+			Cf:       "default",
+			Limit:    MaxScanCntLimit,
+		})
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if resp.GetRegionError() != nil {
+			return errors.New(resp.GetRegionError().String())
+		}
+		iterKvPairs := resp.GetKvs()
+		if !firstLoop { // first is duplicated
+			iterKvPairs = iterKvPairs[1:]
+		}
+		for _, kvpair := range iterKvPairs {
+			if kvpair.Error != nil {
+				return errors.New(kvpair.Error.String())
+			}
+			newKey := utils.FormatAPIV2Key(kvpair.Key, false)
+			// keep the same with tikv-server: https://docs.rs/crc64fast/latest/crc64fast/
+			digest.Reset()
+			digest.Write(newKey)
+			digest.Write(kvpair.Value)
+			checksum.Crc64Xor ^= digest.Sum64()
+			checksum.TotalKvs += 1
+			checksum.TotalBytes += (uint64)(len(newKey) + len(kvpair.Value))
+		}
+		kvCnt := len(iterKvPairs)
+		if kvCnt == 0 {
+			break
+		}
+		curStart = iterKvPairs[kvCnt-1].Key
+		firstLoop = false
+	}
+
+	progressCallBack(backup.RegionUnit)
+	exec.mutex.Lock() // lock for updating checksum
+	defer exec.mutex.Unlock()
+	finalChecksum.UpdateChecksum(checksum.Crc64Xor, checksum.TotalKvs, checksum.TotalBytes)
+	logutil.CL(ctx).Info("region checksum finish", zap.Uint64("region id", regionInfo.Region.Id),
+		logutil.Key("region start", regionInfo.Region.StartKey),
+		logutil.Key("region end", regionInfo.Region.EndKey),
+		zap.Reflect("checksum", finalChecksum))
+	return nil
+}
+
+func (exec *Executor) execChecksumRequest(
+	ctx context.Context,
+	method StorageChecksumMethod,
+	progressCallBack func(backup.ProgressUnit),
+) (Checksum, error) {
+	// reuse restore split codes, but do nothing with split.
+	splitClient := restore.NewSplitClient(exec.pdClient, nil, true)
+	regionInfos, err := restore.PaginateScanRegion(ctx, splitClient, exec.startKey, exec.endKey, restore.ScanRegionPaginationLimit)
+	if err != nil {
+		return Checksum{}, errors.Trace(err)
+	}
+	// regionInfos maybe duplicated
```

Comment:
Why ?

Reply:
I see this comment in somewhere else, but have not see it again. I add a check deduplicte function `checkDuplicateRegion` to debug the failure if it happens.